# Extracting vehicle patent

## 1. 사전 작업
### Open zipfile 
zip file read

In [8]:
import pandas as pd
import zipfile
import csv

zf_patent1 = zipfile.ZipFile('J:/data/patent/harvard_dataset_final_merged_output_1.zip') 
csv_patent1 = zf_patent1.open('harvard_dataset_final_merged_output_1.csv')

csv_patent_reader1 = csv.reader(csv_patent1)

### Where is info about class?
어떤 colummn이 class 정보를 담고 있는지 파악

일단 첫 번째 row의 value부터 살펴보자

In [13]:
header = csv_patent_reader1.next()
row1 = csv_patent_reader1.next()
for i in range(0,len(row1)):
    if row1[i] != '':
        print i, header[i], row1[i]

0  4
1 patent_id_0 3858245
2 project_id 3858245
3 uuid_0 NaN
7635 project_id 3858245
7636 uuid_0 NaN
19492 project_id 3858245
19493 uuid_0 9C
19494 patent_id_0 3858245
19495 assignee_id_0 c714aa408398a69feed41b2f8aee33ed
19496 rawlocation_id_0 morrisplains|nj|
19497 type_0 2.0
19498 name_first_0 nan
19499 name_last_0 nan
19500 organization_0 HAIR AGAIN LTD.
19501 residence_0 nan
19502 nationality_0 nan
19503 sequence_0 0.0
19844 project_id 3858245
19845 uuid_0 1757
19846 patent_id_0 3858245
19847 inventor_id_0 3858245-1
19848 rawlocation_id_0 new york|ny|us
19849 name_first_0 MICHAEL A
19850 name_last_0 NATE II
19851 sequence_0 1
19852 uuid_1 669
19853 patent_id_1 3858245
19854 inventor_id_1 3858245-2
19855 rawlocation_id_1 new york|ny|us
19856 name_first_1 MAURICE A
19857 name_last_1 MANN
19858 sequence_1 2
20377 project_id 3858245
20378 uuid_0 NaN
20402 project_id 3858245
20403 uuid_0 NaN
32093 project_id 3858245
32094 uuid_0 20
32095 patent_id_0 3858245
32096 citation_id_0 3553737
3

class 정보를 담고 있는 column이 몇 번째 column인지 살펴보자, 일단 62457부터 class 정보가 존재한다

In [10]:
for i in range(62457,len(header)+1):
    if 'class' in header[i]:
        print i, header[i]

62457 mainclass_id_0
62458 subclass_id_0
62462 mainclass_id_1
62463 subclass_id_1
62467 mainclass_id_2
62468 subclass_id_2
62472 mainclass_id_3
62473 subclass_id_3
62477 mainclass_id_4
62478 subclass_id_4
62482 mainclass_id_5
62483 subclass_id_5
62487 mainclass_id_6
62488 subclass_id_6
62492 mainclass_id_7
62493 subclass_id_7
62497 mainclass_id_8
62498 subclass_id_8
62502 mainclass_id_9
62503 subclass_id_9
62507 mainclass_id_10
62508 subclass_id_10
62512 mainclass_id_11
62513 subclass_id_11
62517 mainclass_id_12
62518 subclass_id_12
62522 mainclass_id_13
62523 subclass_id_13
62527 mainclass_id_14
62528 subclass_id_14
62532 mainclass_id_15
62533 subclass_id_15
62537 mainclass_id_16
62538 subclass_id_16
62542 mainclass_id_17
62543 subclass_id_17
62547 mainclass_id_18
62548 subclass_id_18
62552 mainclass_id_19
62553 subclass_id_19
62557 mainclass_id_20
62558 subclass_id_20
62562 mainclass_id_21
62563 subclass_id_21
62567 mainclass_id_22
62568 subclass_id_22
62572 mainclass_id_23
62573 sub

IndexError: list index out of range

그 결과, 62457 ~ 63353이 class 정보를 가진 column임을 알 수 있다

하지만 class가 숫자로 분류되어있다. 일단 10개정도의 row만 살펴보면서 어떤 숫자들이 있는지 알아보자

In [17]:
for i in range(1,11):
    row=csv_patent_reader1.next()
    for j in range(62457,63354):
        if 'class' in header[j] and row[j]!='':
            print header[j], row[j]

mainclass_id_0 5
subclass_id_0 59.1
mainclass_id_1 5
subclass_id_1 45
mainclass_id_0 5
subclass_id_0 496
mainclass_id_1 297
subclass_id_1 225
mainclass_id_0 5
subclass_id_0 636
mainclass_id_0 81
subclass_id_0 315
mainclass_id_1 7
subclass_id_1 148
mainclass_id_2 7
subclass_id_2 129
mainclass_id_3 7
subclass_id_3 165
mainclass_id_4 7
subclass_id_4 142
mainclass_id_5 81
subclass_id_5 324
mainclass_id_6 81
subclass_id_6 437
mainclass_id_0 8
subclass_id_0 694
mainclass_id_1 8
subclass_id_1 659
mainclass_id_2 8
subclass_id_2 662
mainclass_id_0 114
subclass_id_0 344
mainclass_id_0 470
subclass_id_0 63
mainclass_id_1 470
subclass_id_1 205
mainclass_id_0 470
subclass_id_0 19
mainclass_id_1 470
subclass_id_1 21
mainclass_id_0 12
subclass_id_0 1A
mainclass_id_0 14
subclass_id_0 71.3


### 질문

1. 왜 patent 1개 (row 1개)에 mainclass, subclass가 최대 179개까지 존재하는가?
 - row 1개의 의미는? : patent 1개, class는 여러 개 가질 수 있음
 - main class만 기준으로 삼아도 되는가? : main class가 3digit level -> 분류하기 적당함
 - 숫자 외에 혹시 문자가 섞여있는 class가 있는가? :
 
2. class 숫자의 의미는?

https://www.uspto.gov/web/patents/classification/selectnumwithtitle.htm 
https://www.uspto.gov/web/patents/classification/international/ipc/ipc8/ipc_concordance/ipcsel.htm
자율주행 :  http://www.slideshare.net/LexInnova/driverless-car-technology-patent-landscape-analysis

44	Fuel and related compositions

70	Locks

73	Measuring and testing

74	Machine element or mechanism

123	Internal-combustion engines

152	Resilient tires and wheels

173	Tool driving or impacting

180	Motor vehicles

188	Brakes

191	Electricity: transmission to vehicles

192	clutches and power-stop control

248	Supports

278 Land vehicles: animal draft appliances

280	Land vehicles

293	Vehicle fenders

294	Handling: hand and hoist-line implements

296	Land vehicles: bodies and tops

297	Chairs and seats

298	Land vehicles: dumping

301	Land vehicles: wheels and axles

303	Fluid-pressure and analogous brake systems

305	Wheel substitutes for land vehicles

307	Electrical transmission or interconnection systems

310	Electrical generator or motor structure

315	Electric lamp and discharge devices: systems

318	Electricity: motive power systems

320	Electricity: battery or capacitor charging or discharging

322	Electricity: single generator systems

323	Electricity: power supply or regulation systems

324	Electricity: measuring and testing

326	Electronic digital logic circuitry

327	Miscellaneous active electrical nonlinear devices, circuits, and systems

340	Communications: electrical

341	Coded data generation or conversion

346	Recorders

361	Electricity: electrical systems and devices

363	Electric power conversion systems

368	Horology: time measuring systems or devices

370	Multiplex communications

375	Pulse or digital communications

377	Electrical pulse counters, pulse dividers, or shift registers: circuits and systems

388	Electricity: motor control systems

403	Joints and connections

429	Chemistry: electrical current producing apparatus, product, and process

439	Electrical connectors

474	Endless belt power transmission systems or components

476	Friction gear transmission systems or components

477	Interrelated power delivery controls, including engine control

492	Roll or roller

700	Data processing: generic control systems or specific applications

701	Data processing: vehicles, navigation, and relative location

702	Data processing: measuring, calibrating, or testing

706	Data processing: artificial intelligence

707	Data processing: database and file management or data structures

709	Electrical computers and digital processing systems: multicomputer data transferring

710	Electrical computers and digital data processing systems: input/output

711	Electrical computers and digital processing systems: memory

712	Electrical computers and digital processing systems: processing architectures and instruction processing (e.g., processors)

713	Electrical computers and digital processing systems: support

714	Error detection/correction and fault detection/recovery

718	Electrical computers and digital processing systems: virtual machine task or process management or task management/control

719	Electrical computers and digital processing systems: interprogram communication or interprocess communication (ipc)

726	Information security

901	Robots

903	Hybrid electric vehicles (hevs)

## 2. Extracting matched rows

### making class list
making list with the numbers of class column

In [1]:
import pandas as pd
import zipfile
import csv

zf_patent1 = zipfile.ZipFile('E:/patent/all_patent/harvard_dataset_final_merged_output_1.zip') 
csv_patent1 = zf_patent1.open('harvard_dataset_final_merged_output_1.csv')
csv_patent_reader1 = csv.reader(csv_patent1)

header = csv_patent_reader1.next()
mainclass_col = []

for j in range(62457,63354):
    if 'mainclass' in header[j]:   
        mainclass_col.append(j)

making csv file with vehicle class

In [3]:
classinfo = pd.read_csv('E:/patent/veh_class.csv')
veh_class = classinfo['class'].tolist()
veh_class = map(int,veh_class)

In [5]:
myfile1 = open("harvard_vehicle_patent1.csv", 'wb')
myfile_writer1 = csv.writer(myfile1,quoting=csv.QUOTE_ALL)
myfile_writer1.writerow(header)

for line in csv_patent_reader1:
    for i in mainclass_col:
        if line[i]=='':
                break
                
        else:
            try:            
                if int(line[i]) in veh_class:
                    myfile_writer1.writerow(line)
                    break
                
            except ValueError:
                pass                
        
       
myfile1.close()

In [7]:
zf_patent2 = zipfile.ZipFile('E:/patent/all_patent/harvard_dataset_final_merged_output_2.zip')
csv_patent2 = zf_patent2.open('harvard_dataset_final_merged_output_2.csv')
csv_patent_reader2 = csv.reader(csv_patent2)

csv_patent_reader2.next()
myfile2 = open("harvard_vehicle_patent2.csv", 'wb')
myfile_writer2 = csv.writer(myfile2,quoting=csv.QUOTE_ALL)
myfile_writer2.writerow(header)

for line in csv_patent_reader2:
    for i in mainclass_col:
        if line[i]=='':
                break
                
        else:
            try:            
                if int(line[i]) in veh_class:
                    myfile_writer2.writerow(line)
                    break
                
            except ValueError:
                pass                
        
       
myfile2.close()

How many cols does csv have?

In [11]:
zf_vh_ptnt1 = zipfile.ZipFile('E:/patent/harvard_vehicle_patent1.zip') 
vh_csv1 = zf_vh_ptnt1.open('harvard_vehicle_patent1.csv')
vh_csv_reader1 = csv.reader(vh_csv1)

header = vh_csv_reader1.next()

n_col=0
for line in vh_csv_reader1:
    n_col=n_col+1

print n_col

519256
